tools/packaging/common/envoy_bootstrap.json

```diff
@@ -1,6 +1,10 @@
 {
   "node": {
+    {{- if .envoy_node_id_template }}
+    "id": "{{ .envoy_node_id_template }}",
```

Comment:
Is this actually a template or just a string? If it's a constant string, we should call it that way.

There are a couple of more places where node ID is used:
- startup flag
- tracing span value
- control plane definitely parses it

Can you make sure those places are also addressed, and we have an end-to-end test exercising this option?

Reply:
* Ah.. It is a complicated a bit. "ISTIO_META_ENVOY_NODE_ID_TEMPLATE" is the template and actually the `.envoy_node_id_template` is a string (but not constant) rendered from "ISTIO_META_ENVOY_NODE_ID_TEMPLATE".  If you have any good name for this, please let me know.
* I don't think it is very necessary to apply for the other places because this is basically for sending a traffic to out of the cluster & non-standard Istiod. But, if there are some needs for this, I can make a following PR.
* RE: testing, I will append it in this PR.

BTW, if there is any concern regarding security or other issue, please let me know as well.

Reply:
`.envoy_node_id` is more appropriate IMHO, with TEMPLATE being the env variable.
I agree the value is not important, but it's hard to tell if anything else makes an assumption on it - best to run a CI test with a custom node to make sure nothing else sets/expects a different value.

Reply:
Actually, this PR has the same idea with https://github.com/istio/istio/blob/71b1c414b5657e2091539375b39ffc2b5af0b8ad/tools/packaging/common/gcp_envoy_bootstrap.json#L4-L8. 
The custom Node ID is only used in xDS requests (+ LRS as well) by this way, and for the other usage of Node Id, still they use the standard Node Id. 

In addition, the customization is assuming the non-standard Istiod implementation.
I don't think that there are needs of customization when `istiod` is used. (Istiod is not able to have multi-tenant.)

WDYT?

Reply:
QQ: it would be helpful to have the extra cluster information (by-customizing) on Node Id in the multi-cluster env?

Reply:
There's no value in encoding in node ID - bootstrap node metadata is a better structured form and passed to every xDS request.

My concern is to prevent reliance on the form of node ID altogether. Ideally, xds proxy and everything else accept a randomly generated node ID, but I don't know how to put a test to avoid regressing in this aspect.

Reply:
> There's no value in encoding in node ID - bootstrap node metadata is a better structured form and passed to every xDS request.

I agree basically, but in our case, encoding on the metadata seems to be not an option because the node metadata requires the extra step of parsing and indexing to get the value.

> My concern is to prevent reliance on the form of node ID altogether. Ideally, xds proxy and everything else accept a randomly generated node ID, but I don't know how to put a test to avoid regressing in this aspect.

Hmm, this seems to be good point for keeping the code resilient. 
Will think about more. Thanks!

Reply:
> I agree basically, but in our case, encoding on the metadata seems to be not an option because the node metadata requires the extra step of parsing and indexing to get the value.

It's encoded either way - either as an ad-hoc text format in node ID, or as a well-designed protobuf encoding as struct in node metadata. I'd be surprised if the standard proto parser is too slow, and even then you can implement a SAX parser over bytes in C++.